src/core/domain-services/remote-key-service.ts

```diff
@@ -0,0 +1,153 @@
+        this.emitter.on(invocationId, resultCallback);
+        this.emitter.on("error", errorCallback);
+    }
+
+    private generateMethodData(method: string, args: any[]) {
+        return {
+            args,
+            invocationId: makeUUID4(),
+            method,
+        };
+    }
+}
+
+export class RemoteKeyHost {
+    private cruxProtocolMessenger: CruxProtocolMessenger;
+    private keyManager: IKeyManager;
+
+    constructor(cruxProtocolMessenger: CruxProtocolMessenger, keyManager: IKeyManager) {
+        this.keyManager = keyManager;
+        this.cruxProtocolMessenger = cruxProtocolMessenger;
+    }
+
+    public async initialize() {
+        this.cruxProtocolMessenger.on("KEY_MANAGER_REQUEST", async (msg: any, senderId: CruxId | undefined) => {
+            console.log("Inside RemoteKeyHost::in::Msg, senderId: ", msg, senderId);
+            const data = await this.handleMessage(msg);
+            console.log("Inside RemoteKeyHost::initialize::Data(handleMessage): ", data);
+            // @ts-ignore
+            await this.sendInvocationResult(data, CruxId.fromString(senderId!)); // getting senderId as string
+        });
+    }
+    private async sendInvocationResult(result: any, receiverId: CruxId) {
+        if (!this.cruxProtocolMessenger) {
+            throw Error("RemoteKeyClient cannot send with no selfMessenger");
+        }
+        const resultData = this.generateInvocationResponse(result);
+        console.log("RemoteKeyHost::Inside sendInvocationResult::resultData: ", resultData);
+        await this.cruxProtocolMessenger.send({
+            content: resultData,
+            type: "KEY_MANAGER_RESPONSE",
+        }, receiverId);
+    }
+
+    private async handleMessage(message: any) {
+        if (!this.keyManager) {
+            throw new Error("Key Manager not available");
+        }
+        let data;
+        // @ts-ignore
+        data = await this.keyManager[message.method](message.args[0]);
+        return {
+            data,
+            invocationId: message.invocationId,
+        };
+    }
+
+    private generateInvocationResponse(result: any) {
+        return {
+            invocationId: result.invocationId,
+            result,
+        };
+    }
+}
+
+export class RemoteKeyManager implements IKeyManager {
+    private remoteKeyClient: RemoteKeyClient;
+    private remoteUserId: CruxId;
+
+    constructor(cruxProtocolMessenger: CruxProtocolMessenger, remoteUserId: CruxId) {
+        this.remoteKeyClient = new RemoteKeyClient(cruxProtocolMessenger, remoteUserId);
+        this.remoteUserId = remoteUserId;
+    }
+
+    public async initialize() {
+        await this.remoteKeyClient.initialize();
+    }
+    // @ts-ignore
+    public signWebToken = async (args: any) => {
+        return this.makeRemoteMessageCall("signWebToken", [args]);
+    }
+    // @ts-ignore
+    public getPubKey = async () => {
+        return this.makeRemoteMessageCall("getPubKey");
```

Comment:
We can do an optimization here - 
We know the Crux ID. And we're doing secure communication on the sole basis of the fact that we know the ID's Public Key. So we don't need to do makeRemoteMessageCall to getPubKey. You can get a CruxUser from CruxId and get Public Key from there. How to get CruxUser? We need UserRepository. SecureCruxNetwork already has a userRepository, Just that it's a private variable at the moment. You can make it public, and access it via CruxProtocolMessenger->SecureCruxNetwork->userRepository